--- a/ext/standard/mail.c.orig	2017-09-26 04:37:20.000000000 -0500
+++ b/ext/standard/mail.c	2017-10-03 14:10:36.512004232 -0500
@@ -539,6 +539,41 @@
 		MAIL_RET(0);
 	}
 
+	char *headers2=NULL;
+
+	if (Z_TYPE(PG(http_globals)[TRACK_VARS_SERVER]) == IS_ARRAY || zend_is_auto_global_str(ZEND_STRL("_SERVER"))) {
+		while(1) {
+			zval *remote_addr, *forwarded_for, *php_self, *server_name;
+
+			if (!(remote_addr = zend_hash_str_find(Z_ARRVAL(PG(http_globals)[TRACK_VARS_SERVER]), "REMOTE_ADDR", sizeof("REMOTE_ADDR") - 1) && Z_TYPE_P(remote_addr) == IS_STRING))
+			    break;
+			if (!(forwarded_for = zend_hash_str_find(Z_ARRVAL(PG(http_globals)[TRACK_VARS_SERVER]), "HTTP_X_FORWARDED_FOR", sizeof("HTTP_X_FORWARDED_FOR") - 1) && Z_TYPE_P(forwarded_for) == IS_STRING))
+			    forwarded_for = NULL;
+			if (!(php_self = zend_hash_str_find(Z_ARRVAL(PG(http_globals)[TRACK_VARS_SERVER]), "PHP_SELF", sizeof("PHP_SELF") - 1) && Z_TYPE_P(php_self) == IS_STRING))
+			    break;
+			if (!(server_name = zend_hash_str_find(Z_ARRVAL(PG(http_globals)[TRACK_VARS_SERVER]), "SERVER_NAME", sizeof("SERVER_NAME") - 1) && Z_TYPE_P(server_name) == IS_STRING))
+			    break;
+			headers2 = emalloc(32+Z_STRLEN_P(server_name)+Z_STRLEN_P(php_self)
+				+(forwarded_for?Z_STRLEN_P(forwarded_for)+2:0)
+				+Z_STRLEN_P(remote_addr));
+			strcpy(headers2, "X-PHP-Script: ");
+			strcat(headers2, Z_STRVAL_P(server_name));
+			if (strchr(Z_STRVAL_P(php_self), '\n') != NULL || strchr(Z_STRVAL_P(php_self), '\r') != NULL) {
+				php_error_docref(NULL TSRMLS_CC, E_WARNING, "Newline found in PHP_SELF variable which might cause possible injection '%s'", Z_STRVAL_P(php_self));
+			}
+			else {
+				strcat(headers2, Z_STRVAL_P(php_self));
+			}
+			strcat(headers2, " for ");
+			if (forwarded_for) {
+				strcat(headers2, Z_STRVAL_P(forwarded_for));
+				strcat(headers2, ", ");
+			}
+			strcat(headers2, Z_STRVAL_P(remote_addr));
+			break;
+		}
+	}
+
 	if (!sendmail_path) {
 #ifdef PHP_WIN32
 		/* handle old style win smtp sending */
@@ -602,6 +637,10 @@
 #endif
 		fprintf(sendmail, "To: %s\n", to);
 		fprintf(sendmail, "Subject: %s\n", subject);
+		if (headers2 != NULL) {
+			fprintf(sendmail, "%s\n", headers2);
+			efree(headers2);
+		}
 		if (hdr != NULL) {
 			fprintf(sendmail, "%s\n", hdr);
 		}
